test/blackbox-tests/test-cases/toplevel-integration/run.t

```diff
@@ -0,0 +1,9 @@
+Test toplevel-init-file on empty project
+----------------------------------------------------
+  $ dune toplevel-init-file
+  Info: Creating file dune-project with this contents:
+  | (lang dune 2.1)
+  | (name test)
+  #directory "$TESTCASE_ROOT/_build/default/.test.objs/byte";;
+  #directory "$TESTCASE_ROOT/_build/default/.test.objs/native";;
+  #load "$TESTCASE_ROOT/_build/default/test.cma";;
```

Comment:
Could you add a test testing the whole story? i.e. doing `#use "..."` from a toplevel? And also one testing an error case, just to see what the error looks like. You can do the following:

```
  $ ocaml -stdin <<EOF
  > #use "..."
  > ...use code that was just loaded...
  > EOF
```

Reply:
Okay, I have several questions here:
1. Even outside of tests, doing just `#use "dune.top"` doesn't work in `ocaml`(it can't find the file), but it does in `utop`. Is `ocaml` using `OCAML_TOPLEVEL_PATH` or does it resolve paths in some other fashion? I wasn't able to figure this out by googling and I only use plain `ocaml` rarely. But of course, we want to support it, that's the whole point of this PR! :)
2. When running the tests, `OCAML_TOPLEVEL_PATH` still points to my user-local directory. Am I supposed to run `export OCAML_TOPLEVEL_PATH="_build/install/default/lib/toplevel/dune.top"` or is there a nicer way to refer to the local installation we're testing?
3. What error case do you have in mind?

Reply:
I did a bit of research and right now, packages that install toplevel scripts such as `ocamlfind` or `down` install them in both `<libdir>/toplevel` and `<libdir>/ocaml`. The latter is immediately available in the toplevel.

`OCAML_TOPLEVEL_PATH` is only used by utop and not by the vanilla toplevel. However, since 4.08.0, there is a new variable `OCAMLTOP_INCLUDE_PATH` that is supported natively. I opened a ticket to discuss setting this variable in opam: https://github.com/ocaml/opam-repository/issues/15547. Let's see what happens there. At worse we could install `dune.top` in `<libdir>/ocaml` as well.

On the dune side, dune already extends a few environment variable so that when running dune or other tools inside dune, various tools will look for things in `_build/install` in addition to things installed globally. We should do this for `OCAMLTOP_INCLUDE_PATH` as well. More precisely, you can look for `extend_var "OCAMLPATH"` in `src/dune/context.ml` and add the following line: `extend_var "OCAMLTOP_INCLUDE_PATH" (Path.relative local_lib_path "toplevel")`.

Then, in the test `#use "dune.top"` will work out of the box with OCaml >= 4.08. We run the testsuite with 4.09 in the CI so having the test only work with OCaml >= 4.08 is fine.

Regarding the error case, I was thinking of a case where compilation fails. For instance, what does the user see if one of the ml file that `dune` tries to compile doesn't compile? Or if a `dune` file is invalid.

Reply:
Thanks for all this great info and sorry about my (usual) late response!

Here's what I've done for now. Since there are still several unknowns, I've also had to make several questionable choices.

1. I've added `extend_var` as you suggested.
2. I've also installed `dune.top` into `<libdir>/ocaml`, since we can't rely on the opam PR being finished anytime soon (right?).
3. I've put back the documentation bit that people need to do `#use "topfind"` first. Indeed, it turns out that without that I _don't_ have `Toplevel` module available in `dune.top` (I missed this because I had `#use "topfind"` in my `.ocamlinit`). Do you think it's possible to remove the dependency on `#use "topfind"` or, even better, on the `Toplevel` module?
4. I've added a few tests, including failures. However, the failures expose compilation internals and I'm not sure whether the tests won't break randomly: https://github.com/mbernat/dune/blob/toplevel/test/blackbox-tests/test-cases/toplevel-integration/run.t#L35

Please let me know what you think about this.

In case I don't get around to this again for a week, I wish you a very Merry Christmas and thanks again for all your help! 🎄

Reply:
Looking at those errors more closely, they actually reference my local paths, so they're definitely wrong.

How do you handle test errors containing local paths? And more importantly, what kind of error behavior should we expect from running `dune toplevel-init-file` and `dune.top`?

Reply:
Just back from holidays. Thanks, hope you had a nice Christmas too!

For 2, that seems reasonable.

For 3, we can do the same as the `topfind` script, i.e. temporary add the `+compiler-libs` directory:

```ocaml
#directory "+compiler-libs";;
...
#remove_directory "+compiler-libs";;
```

For 4, I was surprised by this output as normally Dune doesn't print the full compilation command line if it detects that the error messages already points to a file, i.e. is of the form `File "xxx", ...`. After a bit of digging, I found that we always print the full compilation command when dune is run inside dune because of this in `src/dune/config.ml`:

```ocaml
let show_full_command_on_error () =
  inside_dune || inside_ci || !Clflags.always_show_command_line
```

@rgrinberg, I don't remember the details. Was the `inside_dune ||` part intentional?